File: windows/src/core/sample-clock.js
// Smooths audio buffer timestamps with a second-order delay-locked loop, so decoded LTC
// frames get stable times even when buffers arrive in bursts with jittery timestamps.
// JavaScript port of Sources/Core/SampleClock.swift. Times are in milliseconds.

export class SampleClock {
  constructor(sampleRate) {
    this.nominalMsPerSample = 1000 / sampleRate;
    this.msPerSample = this.nominalMsPerSample;
    this.predicted = null;   // expected time of the next buffer's first sample
    this.resetThreshold = 50;
  }

  /** Filtered time of the first sample of a buffer whose raw timestamp is `observed`. */
  timestamp(observed, frames) {
    if (frames <= 0) return observed;
    if (this.predicted == null || Math.abs(observed - this.predicted) >= this.resetThreshold) {
      // First buffer, or a real discontinuity (dropout, device restart).
      this.msPerSample = this.nominalMsPerSample;
      this.predicted = observed + frames * this.msPerSample;
      return observed;
    }
    // Loop bandwidth ~0.5 Hz: jitter is averaged over a couple of seconds.
    const periodSeconds = (frames * this.msPerSample) / 1000;
    const omega = 2 * Math.PI * 0.5 * periodSeconds;
    const b = Math.SQRT2 * omega, c = omega * omega;
    const error = observed - this.predicted;
    const filtered = this.predicted + b * error;
    this.msPerSample += (c * error) / frames;
    this.msPerSample = Math.min(Math.max(this.msPerSample, this.nominalMsPerSample * 0.995), this.nominalMsPerSample * 1.005);
    this.predicted = filtered + frames * this.msPerSample;
    return filtered;
  }
}

File: Sources/Core/SampleClock.swift
// Smooths audio buffer timestamps. Some drivers (notably virtual/network
// soundcards such as Dante Virtual Soundcard) deliver buffers in bursts with
// jittery timestamps; this second-order delay-locked loop tracks the true
// sample clock so decoded LTC frames get stable times.

final class SampleClock {
    private let nominalTicksPerSample: Double
    private let ticksPerSecond: Double
    private var ticksPerSample: Double
    private var predicted: Double?   // expected time of the next buffer's first sample
    private let resetThreshold: Double

    init(sampleRate: Double, ticksPerSecond: Double) {
        self.ticksPerSecond = ticksPerSecond
        nominalTicksPerSample = ticksPerSecond / sampleRate
        ticksPerSample = nominalTicksPerSample
        resetThreshold = 0.05 * ticksPerSecond
    }

    var currentTicksPerSample: Double { ticksPerSample }

    /// Returns the filtered time of the first sample of a buffer whose raw
    /// timestamp is `observed`, and advances by `frames` samples.
    func timestamp(observed: Double, frames: Int) -> Double {
        guard frames > 0 else { return observed }
        guard let p = predicted, abs(observed - p) < resetThreshold else {
            // First buffer, or a real discontinuity (dropout, device restart).
            ticksPerSample = nominalTicksPerSample
            predicted = observed + Double(frames) * ticksPerSample
            return observed
        }
        // Loop bandwidth ~0.5 Hz: jitter is averaged over a couple of seconds.
        let period = Double(frames) * ticksPerSample / ticksPerSecond
        let omega = 2 * Double.pi * 0.5 * period
        let b = 2.0.squareRoot() * omega
        let c = omega * omega

        let error = observed - p
        let filtered = p + b * error
        ticksPerSample += c * error / Double(frames)
        // Clocks never differ by more than a fraction of a percent.
        ticksPerSample = min(max(ticksPerSample, nominalTicksPerSample * 0.995), nominalTicksPerSample * 1.005)
        predicted = filtered + Double(frames) * ticksPerSample
        return filtered
    }
}
